pyro/infer/reparam/neutra.py

```diff
@@ -0,0 +1,66 @@
+import pyro
+import pyro.distributions as dist
+from pyro.infer.autoguide.guides import AutoContinuous
+
+from .reparam import Reparam
+
+
+class NeuTraReparam(Reparam):
+    """
+    Neural Transport reparameterizer [1] of multiple latent variables.
+
+    This uses a trained :class:`~pyro.infer.autoguide.AutoContinuous`
+    guide to alter the geometry of a model, typically for use e.g. in MCMC.
+    Example usage::
+
+        # Step 1. Train a guide
+        guide = AutoIAFNormal(model)
+        svi = SVI(model, guide, ...)
+        # ...train the guide...
+
+        # Step 2. Use trained guide in NeuTra MCMC
+        neutra = NeuTraReparam(guide)
+        model = poutine.infer_config(lambda _: {"reparam": neutra})
+        model = poutine.reparam(model)
+        nuts = NUTS(model)
+        # ...now use the model in HMC or NUTS...
+
+    Note that all sites must share a single common :class:`NeuTraReparam` instance.
+
+    [1] Hoffman, M. et al. (2019)
+        "NeuTra-lizing Bad Geometry in Hamiltonian Monte Carlo Using Neural Transport"
+        https://arxiv.org/abs/1903.03704
+
+    :param ~pyro.infer.autoguide.AutoContinuous guide: A trained guide.
+    """
+    def __init__(self, guide):
+        if not isinstance(guide, AutoContinuous):
+            raise TypeError("NeuTraReparam expected an AutoContinuous guide, but got {}"
+                            .format(type(guide)))
+        self.guide
+        self.x_constrained = []
+
+    def __call__(self, name, fn, obs):
+        log_density = 0.
```

Comment:
IIUC correctly, if we are to use this directly with HMC/NUTS, we will also probably need a `sample` statement corresponding to the `obs`, maybe something like 

```python
if obs is not None:
    return fn, obs
```

Reply:
Good point. I've added the following early exit:
```py
if name not in self.guide.prototype_trace.nodes:
    return fn, obs
```
Note that most reparameterizers do not support observation statements, and so I've asserted
```py
assert obs is None, "ThisReparam does not support observe statements"
```
The reason I'm opting to error rather than be permissive is to address the mismatch of behavior between the `obs=___` kwarg and `poutine.condition`. Due to poutine stack ordering, reparameterizers cannot know about `poutine.condition`s higher in the stack, but they do know about `obs=___` kwargs. For most reparameterizers `poutine.condition` will result in infinite loss, which is an easy-to-identify problem. Because the `poutine.condition` case is problematic, I have chosen to make the `obs=___` case also problematic (raising `AssertionError`), so that if users switch between the two idioms they will see identical `Reparam` behavior (i.e. either both good or both bad).

Reply:
Thanks for explaining. We use [poutine.condition](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/mcmc/util.py#L257) to compute the potential energy for samples from the integrator. Will that be problematic? 

```python
        cond_model = poutine.condition(self.model, params_constrained)
        model_trace = poutine.trace(cond_model).get_trace(*self.model_args,
                                                          **self.model_kwargs)
```

Reply:
It should be fine for MCMC to use of `poutine.condition` to set latent variables, since those latent variables will be uneffected by `poutine.reparam` statements (they are either not reparameterized sites, or they are auxiliary sites created by `poutine.reparam` and themselves are not reparameterized).

The only issue is when `poutine.condition` is used for likelihoods: reparameterization often results in a `Delta` simulating a `pyro.deterministic` and would lead to infinite loss when using `poutine.condition`. Therefore most reparameterizers do not support likelihoods (exceptions being `SymmetricStableReparam` and `StableHMMReparam` which produce `Normal`s rather than `Delta`s).

Reply:
Thanks, that makes sense!